src/encoded/static/libs/bootstrap/tooltip.js

```diff
@@ -0,0 +1,120 @@
+//     <TooltipContent />
+// </Tooltip>
+//
+// The component that the user can hover over to trigger the tooltip is given in the `trigger`
+// property and appears in the DOM whereever you place the <Tooltip> component. Put the contents
+// of the tooltip between the opening and closing <Tooltip> tags. This can simply be a string if
+// that's all you need, or a full-fledged component.
+
+export default class Tooltip extends React.Component {
+    constructor() {
+        super();
+        this.state = {
+            tooltipDisplayed: false,
+        };
+        this.timer = null;
+        this.hoveringOverTooltip = false;
+        this.handleTriggerMouseEnter = this.handleTriggerMouseEnter.bind(this);
+        this.handleTriggerMouseLeave = this.handleTriggerMouseLeave.bind(this);
+        this.handleTooltipMouseEnter = this.handleTooltipMouseEnter.bind(this);
+        this.handleTooltipMouseLeave = this.handleTooltipMouseLeave.bind(this);
+        this.handleTriggerFocus = this.handleTriggerFocus.bind(this);
+        this.handleTriggerBlur = this.handleTriggerBlur.bind(this);
+    }
+
+    handleTriggerMouseEnter() {
+        // Mouse entered the tooltip trigger element.
+        this.setState({ tooltipDisplayed: true });
+
+        // If we happen to have a running timer, clear it so we don't hide the tooltip while
+        // hovering over the trigger.
+        if (this.timer) {
+            clearTimeout(this.timer);
+            this.timer = null;
+        }
+    }
+
+    handleTriggerMouseLeave() {
+        // Start a timer that might hide the tooltip after a second passes. It won't hide the
+        // tooltip if they're now hovering over the tooltip itself.
+        this.timer = setTimeout(() => {
+            this.timer = null;
+            if (!this.hoveringOverTooltip) {
+                this.setState({ tooltipDisplayed: false });
+            }
+        }, 1000);
+    }
+
+    handleTriggerFocus() {
+        // Show tooltip when keyboard focus lands on tooltip trigger.
+        this.setState({ tooltipDisplayed: true });
+    }
+
+    handleTriggerBlur() {
+        // Hide tooltip when keyboard focus leaves tooltip trigger.
+        this.setState({ tooltipDisplayed: false });
+    }
+
+    handleTooltipMouseEnter() {
+        // Mouse started hovering over the tooltip itself. Prevent the tooltip from hiding even if
+        // the one-second trigger time expires.
+        this.hoveringOverTooltip = true;
+    }
+
+    handleTooltipMouseLeave() {
+        // Mouse stopped hovering over the tooltip itself. If the timer's not running, hide the
+        // tooltip, otherwise let the existing timer hide the tooltip when it expires.
+        this.hoveringOverTooltip = false;
+        if (!this.timer) {
+            this.setState({ tooltipDisplayed: false });
+        }
+    }
+
+    render() {
+        const { trigger, direction, tooltipId, css } = this.props;
+        const tooltipCss = `tooltip ${direction}`;
```

Comment:
Personal Preference Comment: Why is there no defensive programming to default to a direction if an unknown value is provided.

Reply:
defaultProps defines “bottom” as default. propTypes defines the accepted values.

Reply:
@forresttanaka What I mean is that a user can put in "this-should-not-work" as the value of direction. I doubt if you were expecting that. So, the code ought to set a value for cases where a user puts in something he/she/gender-neutral ought not put to put in.

Reply:
If they put “this-should-not-work” in a dev build, they’ll get an error in the console because of PropTypes. defaultProps sets “bottom” as default. That’s the whole point of PropTypes and default: value and type checking.